scripts/test-new-tests.mjs

```diff
@@ -39,16 +39,34 @@ async function main() {
     return
   }
 
-  try {
-    await execa('git remote set-branches --add origin canary', EXECA_OPTS_STDIO)
-    await execa('git fetch origin canary --depth=20', EXECA_OPTS_STDIO)
-  } catch (err) {
-    console.error(await execa('git remote -v', EXECA_OPTS_STDIO))
-    console.error(`Failed to fetch origin/canary`, err)
+  let diffRevision
+  if (process.env.GITHUB_ACTIONS === 'true') {
+    // GH Actions run on the merge commit so HEAD~1:
+    // 1. includes all changes in the PR
+    // 2. Includes all changes of the commit for pushes
+    diffRevision = 'HEAD~1'
```

Comment:
This is only looking back one commit. What if you make two commits locally and then `git push`? This change will only test changes from one of those commits instead of both.

Reply:
Not on GH actions runs since it runs on the merge commit which is what this branch covers. Outside of GH actions we use the old strategy.

Reply:
Can you explain what "runs on the merge commit" means here? Is that when you click the "Update Branch" button or is it when the PR is merged to canary? 

<img width="883" alt="image" src="https://github.com/vercel/next.js/assets/229881/3b3ab957-ce44-428c-bb79-26e3c0766836">

Reply:
> Is that when you click the "Update Branch" button or is it when the PR is merged to canary?

When the GitHub actions job is created which simulates "when the PR is merged to canary" 

It doesn't run on the head commit on the branch (i.e. what you would run the tests on) but on the commit of the merge branch. The GH Actions docs try to explain this:
> Note that GITHUB_SHA for this event is the last merge commit of the pull request merge branch. If you want to get the commit ID for the last commit to the head branch of the pull request, use github.event.pull_request.head.sha instead.

In https://github.com/vercel/next.js/actions/runs/8605523039/job/23581990844?pr=64117#step:27:30 you can see that `HEAD~1` is sufficient. I cherry-picked this PR onto https://github.com/vercel/next.js/pull/64117 and diffing against `HEAD~1` got all changes of the branch that consisted of 4 commits.

Imagine you have this history:
```
A-B-C-main
 \
  D-E-branch
```

By default, GH Actions will not check out the `branch` ref (i.e. the head of the branch). Instead, it will merge `branch` into `main` and then run on that commit:

```
A-B-C-main - F
 \          /
  D-E-branch
```
F is the commit GH actions (i.e. `actions/checkout`) checks out by default.

In that scenario diffing against `HEAD~1` will be sufficient to include all changes that are exclusive to the branch (since it just includes the merge commit) whereas diffing against `main` would include all the changes exclusive to `branch` *and* all the changes on `main` since `branch` was forked. I don't see how including changes in B and C in CI for PRs is helpful.

You can also see this behavior in the logs of `actions/checkout`: https://github.com/vercel/next.js/actions/runs/8605588371/job/23582826423?pr=64218#step:9:93
This is the log for GH Actions of this PR. The head commit is 9e87c4ce2fedcfd8275e9108523ed1e2a0d12df0 i.e. the commit `HEAD` would point to if you'd check out this branch locally. However, in CI, `HEAD` points to `dda5fbd7` which is the merge commit of merging 9e87c4ce2fedcfd8275e9108523ed1e2a0d12df0 into its fork point with `canary` (`f9ad49fc6f57f3bcf4f508b59deab695f06c74b1`): 
> HEAD is now at dda5fbd7 Merge 9e87c4ce2fedcfd8275e9108523ed1e2a0d12df0 into f9ad49fc6f57f3bcf4f508b59deab695f06c74b1

The DefinitelyTyped repo is using the exact same strategy: https://github.com/DefinitelyTyped/DefinitelyTyped/blame/master/scripts/pnpm-install.sh#L22

Reply:
It sounds like we should be comparing against `git merge-base` for PRs instead

https://git-scm.com/docs/git-merge-base

Reply:
We should do that for local runs, yes. I already added a TODO item below for that. I don't expect this script to be run locally that often so I left it undone for now.

This would be equivalent to the changes proposed for runs in GitHub actions already.

Reply:
Does it run when PRs are created or only when merged to main/canary?

Reply:
This script? On every synchronize (every push to the PR branch). But the the diff revision works for PRs created, when running on canary (i.e. when merged) and when we push to a PR branch.

Reply:
Yeah so my worry here is that `HEAD~1` is only testing one commit back but a PR can push multiple commits so there could be untested commits if you only test one of them.